saleor/core/db/connection.py

```diff
@@ -0,0 +1,139 @@
+
+
+@contextmanager
+def allow_writer_in_context(context: SaleorContext):
+    """Context manager that allows write access to the default database connection in a context (SaleorContext).
+
+    This is a helper context manager that conditionally allows write access based on the
+    database connection name in the given context.
+    """
+    conn_name = get_database_connection_name(context)
+    if conn_name == settings.DATABASE_CONNECTION_DEFAULT_NAME:
+        with allow_writer():
+            yield
+    else:
+        yield
+
+
+def restrict_writer_middleware(get_response):
+    """Middleware that restricts write access to the default database connection.
+
+    This middleware will raise an error if a write operation is attempted on the default
+    database connection. To allow writes, use the `allow_writer` context manager. Make
+    sure that writer is not used accidentally and always use the `using` queryset method
+    with proper database connection name.
+    """
+
+    def middleware(request):
+        with connections[writer].execute_wrapper(_restrict_writer):
+            with connections[replica].execute_wrapper(_restrict_writer):
+                return get_response(request)
+
+    return middleware
+
+
+def _restrict_writer(execute, sql, params, many, context):
+    conn: BaseDatabaseWrapper = context["connection"]
+    allow_writer = getattr(conn, "_allow_writer", False)
+    if conn.alias == writer and not allow_writer:
+        raise UnsafeWriterAccessError(
+            f"Unsafe writer DB access, use `allow_writer` context manager. SQL: {sql}"
+        )
+    elif conn.alias == replica:
+        if not _is_read_only_query(sql):
+            raise UnsafeReplicaUsageError(f"Unsafe replica usage. SQL: {sql}")
+    return execute(sql, params, many, context)
+
+
+def _is_read_only_query(sql_query: str) -> bool:
+    for query in sqlparse.parse(sql_query):
+        query_type = query.get_type().upper()
+        if query_type != "SELECT":
+            return False
+    return True
+
+
+def log_writer_usage_middleware(get_response):
+    """Middleware that logs write access to the default database connection.
+
+    This is similar to the `restrict_writer_middleware` middleware, but instead of
+    raising an error, it logs a message when a write operation is attempted on the
+    default database connection.
+    """
+
+    def middleware(request):
+        with connections[writer].execute_wrapper(_log_writer_usage):
+            return get_response(request)
+
+    return middleware
+
+
+def _log_writer_usage(execute, sql, params, many, context):
+    conn: BaseDatabaseWrapper = context["connection"]
+    allow_writer = getattr(conn, "_allow_writer", False)
+    if conn.alias == writer and not allow_writer:
+        stack_trace = traceback.extract_stack(limit=20)
+        error_msg = color_style().NOTICE(
+            "Unsafe writer DB access, use `allow_writer` context manager."
```

Comment:
```suggestion
            "Unsafe access to the writer DB detected. Call `using()` on the `QuerySet` to utilize a replica DB, or employ the `allow_writer` context manager to explicitly permit access to the writer."
```
Just a suggestion to maybe extend this log message. Just wrapping in `allow_writer` is not always a solution.

Reply:
Done